File: vm-agent/src/scripts/mineru_upload.py
#!/usr/bin/env python3
"""Upload local files to MinerU batch API and export main markdown.

This script is intentionally stdlib-only so it can run in constrained runtimes.
"""

from __future__ import annotations

import argparse
import io
import json
import pathlib
import re
import shutil
import sys
import tempfile
import time
import urllib.error
import urllib.request
import zipfile

SUPPORTED_EXTENSIONS = {".pdf", ".docx", ".doc", ".pptx", ".ppt", ".png", ".jpg", ".jpeg"}
DEFAULT_TIMEOUT_SEC = 600
DEFAULT_POLL_INTERVAL_SEC = 2.0


def _log(message: str) -> None:
    print(f"[mineru_upload] {message}", file=sys.stderr, flush=True)


def _emit_stdout(payload: dict) -> None:
    print(json.dumps(payload, ensure_ascii=False), flush=True)


def sanitize_filename(text: str) -> str:
    sanitized = re.sub(r"[^a-zA-Z0-9._-]+", "_", text.strip())
    sanitized = sanitized.strip("._")
    if not sanitized:
        return "document"
    return sanitized[:120]


def _http_json(
    method: str,
    url: str,
    *,
    headers: dict[str, str] | None = None,
    payload: dict | None = None,
    timeout: int = 60,
) -> dict:
    req_headers = {"Accept": "application/json"}
    if headers:
        req_headers.update(headers)

    data = None
    if payload is not None:
        data = json.dumps(payload).encode("utf-8")
        req_headers.setdefault("Content-Type", "application/json")

    req = urllib.request.Request(url=url, data=data, method=method.upper(), headers=req_headers)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            body = resp.read()
            if not body:
                return {}
            return json.loads(body.decode("utf-8", errors="replace"))
    except urllib.error.HTTPError as exc:
        body = ""
        try:
            body = exc.read().decode("utf-8", errors="replace")
        except Exception:
            pass
        raise RuntimeError(f"HTTP {exc.code} {method.upper()} {url}: {body[:800]}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Network error {method.upper()} {url}: {exc}") from exc


def _http_put_file(url: str, file_path: pathlib.Path, *, timeout: int = 300) -> None:
    data = file_path.read_bytes()
    req = urllib.request.Request(url=url, data=data, method="PUT")
    try:
        with urllib.request.urlopen(req, timeout=timeout):
            return
    except urllib.error.HTTPError as exc:
        body = ""
        try:
            body = exc.read().decode("utf-8", errors="replace")
        except Exception:
            pass
        raise RuntimeError(f"HTTP {exc.code} PUT {url}: {body[:800]}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Network error PUT {url}: {exc}") from exc


def _http_bytes(url: str, *, timeout: int = 180) -> bytes:
    req = urllib.request.Request(url=url, method="GET")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return resp.read()
    except urllib.error.HTTPError as exc:
        body = ""
        try:
            body = exc.read().decode("utf-8", errors="replace")
        except Exception:
            pass
        raise RuntimeError(f"HTTP {exc.code} GET {url}: {body[:800]}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Network error GET {url}: {exc}") from exc


def _must_success(response: dict, context: str) -> dict:
    code = response.get("code")
    if code not in (None, 0):
        raise RuntimeError(f"{context} failed: {response}")
    data = response.get("data")
    if isinstance(data, dict):
        return data
    return response


def _first_presigned_url(batch_data: dict) -> str:
    files = batch_data.get("files")
    if not isinstance(files, list) or not files:
        raise RuntimeError(f"Batch response missing files: {batch_data}")

    first = files[0]
    if not isinstance(first, dict):
        raise RuntimeError(f"Batch response files[0] invalid: {batch_data}")

    for key in ("presigned_url", "upload_url", "url"):
        value = first.get(key)
        if isinstance(value, str) and value:
            return value
    raise RuntimeError(f"Batch response missing presigned_url: {first}")


def create_batch(
    *,
    api_base: str,
    token: str,
    file_name: str,
    language: str,
    timeout_sec: int,
) -> tuple[str, str]:
    endpoint = api_base.rstrip("/") + "/api/v4/file-urls/batch"
    payload = {
        "enable_formula": True,
        "enable_table": True,
        "language": language,
        "files": [{"name": file_name, "is_ocr": True}],
    }
    response = _http_json(
        "POST",
        endpoint,
        headers={"Authorization": f"Bearer {token}"},
        payload=payload,
        timeout=min(timeout_sec, 120),
    )
    data = _must_success(response, "create batch")

    batch_id = data.get("batch_id") or response.get("batch_id")
    if not isinstance(batch_id, str) or not batch_id:
        raise RuntimeError(f"Batch response missing batch_id: {response}")

    presigned_url = _first_presigned_url(data)
    return batch_id, presigned_url


def poll_batch(
    *,
    api_base: str,
    token: str,
    batch_id: str,
    timeout_sec: int,
    poll_interval_sec: float = DEFAULT_POLL_INTERVAL_SEC,
) -> tuple[dict, str]:
    endpoint = api_base.rstrip("/") + f"/api/v4/extract-results/batch/{batch_id}"
    start = time.monotonic()
    last_state = None

    while True:
        response = _http_json(
            "GET",
            endpoint,
            headers={"Authorization": f"Bearer {token}"},
            timeout=min(timeout_sec, 60),
        )
        data = _must_success(response, "poll batch")

        state = data.get("state") or response.get("state")
        if isinstance(state, str) and state != last_state:
            elapsed = time.monotonic() - start
            _log(f"state={state} elapsed={elapsed:.1f}s")
            last_state = state

        if state == "done":
            extract_result = data.get("extract_result")
            if extract_result is None:
                extract_result = data.get("extract_results")
            if extract_result is None:
                extract_result = response.get("extract_result")
            if extract_result is None:
                extract_result = response.get("extract_results")

            if isinstance(extract_result, dict):
                items = [extract_result]
            elif isinstance(extract_result, list):
                items = [item for item in extract_result if isinstance(item, dict)]
            else:
                items = []

            if not items:
                raise RuntimeError(f"Batch done but missing extract_result: {response}")

            chosen = next((item for item in items if isinstance(item.get("full_zip_url"), str) and item.get("full_zip_url")), items[0])
            zip_url = chosen.get("full_zip_url")
            if not isinstance(zip_url, str) or not zip_url:
                raise RuntimeError(f"Batch done but full_zip_url missing: {chosen}")
            return chosen, zip_url

        if state == "failed":
            err_msg = data.get("err_msg") or response.get("err_msg") or data.get("message") or response.get("message")
            raise RuntimeError(f"MinerU batch failed: {err_msg or '(no err_msg)'}")

        if time.monotonic() - start > timeout_sec:
            raise RuntimeError(f"Poll timeout after {timeout_sec}s (last state={state})")

        time.sleep(poll_interval_sec)


def _safe_extract_zip(zip_bytes: bytes, out_dir: pathlib.Path) -> list[pathlib.Path]:
    out_dir.mkdir(parents=True, exist_ok=True)
    extracted_files: list[pathlib.Path] = []
    with zipfile.ZipFile(io.BytesIO(zip_bytes)) as zf:
        root = out_dir.resolve()
        for member in zf.infolist():
            target = (out_dir / member.filename).resolve()
            if root not in target.parents and target != root:
                raise RuntimeError(f"Unsafe zip path: {member.filename}")
            zf.extract(member, out_dir)

    for path in out_dir.rglob("*"):
        if path.is_file():
            extracted_files.append(path)
    return extracted_files


def _markdown_score(md_path: pathlib.Path) -> tuple[int, int]:
    name = md_path.name.lower()
    rel = md_path.as_posix().lower()
    penalty = 0
    bonus = 0

    if "readme" in name:
        penalty += 3
    if "layout" in rel or "span" in rel or "debug" in rel:
        penalty += 3
    if "middle" in rel or "origin" in rel:
        penalty += 2
    if "content" in name or "result" in name or "output" in name:
        bonus += 2

    size = md_path.stat().st_size
    return (bonus - penalty, size)


def extract_main_markdown(extracted_files: list[pathlib.Path]) -> pathlib.Path:
    candidates = [p for p in extracted_files if p.suffix.lower() in {".md", ".markdown"}]
    if not candidates:
        raise RuntimeError("No markdown file found in MinerU result zip")
    candidates.sort(key=_markdown_score, reverse=True)
    return candidates[0]


def copy_images(extract_root: pathlib.Path, images_out_dir: pathlib.Path) -> int:
    image_dirs = [p for p in extract_root.rglob("*") if p.is_dir() and p.name.lower() == "images"]
    if not image_dirs:
        return 0

    copied = 0
    images_out_dir.mkdir(parents=True, exist_ok=True)
    for src_dir in image_dirs:
        for src_file in src_dir.rglob("*"):
            if not src_file.is_file():
                continue
            rel = src_file.relative_to(src_dir)
            dst = images_out_dir / rel
            dst.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(src_file, dst)
            copied += 1
    return copied


def run(args: argparse.Namespace) -> dict:
    started = time.monotonic()
    input_path = pathlib.Path(args.file_path).expanduser().resolve()
    output_dir = pathlib.Path(args.output_dir).expanduser().resolve()

    if not input_path.exists() or not input_path.is_file():
        raise RuntimeError(f"Input file not found: {input_path}")

    extension = input_path.suffix.lower()
    if extension not in SUPPORTED_EXTENSIONS:
        supported = ", ".join(sorted(ext.lstrip(".") for ext in SUPPORTED_EXTENSIONS))
        raise RuntimeError(f"Unsupported file format '{extension}'. Supported: {supported}")

    timeout_sec = int(args.timeout)
    if timeout_sec <= 0:
        raise RuntimeError("--timeout must be > 0")

    output_dir.mkdir(parents=True, exist_ok=True)
    _log(f"input={input_path} output_dir={output_dir}")

    _log("creating upload batch")
    batch_id, presigned_url = create_batch(
        api_base=args.api_base,
        token=args.token,
        file_name=input_path.name,
        language=args.language,
        timeout_sec=timeout_sec,
    )
    _log(f"batch_id={batch_id}")

    _log("uploading file to presigned URL")
    _http_put_file(presigned_url, input_path, timeout=min(timeout_sec, 300))

    _log("polling extraction status")
    extract_result, zip_url = poll_batch(
        api_base=args.api_base,
        token=args.token,
        batch_id=batch_id,
        timeout_sec=timeout_sec,
    )

    _log("downloading result zip")
    zip_bytes = _http_bytes(zip_url, timeout=min(timeout_sec, 300))

    with tempfile.TemporaryDirectory(prefix="mineru-upload-") as temp_dir:
        extract_root = pathlib.Path(temp_dir) / sanitize_filename(batch_id)
        extracted_files = _safe_extract_zip(zip_bytes, extract_root)

        main_md = extract_main_markdown(extracted_files)
        markdown_path = output_dir / f"{sanitize_filename(input_path.stem)}.md"
        if markdown_path.exists() and markdown_path.resolve() != main_md.resolve():
            markdown_path = output_dir / f"{sanitize_filename(input_path.stem)}-{int(time.time())}.md"
        shutil.copy2(main_md, markdown_path)

        images_dir = output_dir / "images"
        copied_images = copy_images(extract_root, images_dir)
        if copied_images == 0:
            images_dir = output_dir / "images"

    markdown_text = markdown_path.read_text(encoding="utf-8", errors="replace")
    elapsed = time.monotonic() - started
    _log(f"completed batch_id={batch_id} chars={len(markdown_text)} elapsed={elapsed:.1f}s")
    print(
        json.dumps(
            {
                "batch_id": batch_id,
                "zip_url": zip_url,
                "output_dir": str(output_dir),
                "markdown_path": str(markdown_path),
                "images_dir": str(images_dir),
                "copied_images": copied_images,
                "elapsed_sec": round(elapsed, 3),
            },
            ensure_ascii=False,
        ),
        file=sys.stderr,
    )

    return {
        "ok": True,
        "markdown_path": str(markdown_path),
        "images_dir": str(images_dir),
        "chars": len(markdown_text),
    }


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Upload a local file to MinerU batch API and export markdown.")
    parser.add_argument("file_path", help="Path to local input file")
    parser.add_argument("--token", required=True, help="MinerU API token (Bearer token value)")
    parser.add_argument("--output-dir", required=True, help="Directory to write markdown and images")
    parser.add_argument("--language", default="ch", help="MinerU language hint (default: ch)")
    parser.add_argument("--timeout", type=int, default=DEFAULT_TIMEOUT_SEC, help="Total timeout seconds (default: 600)")
    parser.add_argument("--api-base", default="https://mineru.net", help="MinerU API base URL")
    return parser


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()

    try:
        result = run(args)
        _emit_stdout(result)
        return 0
    except KeyboardInterrupt:
        _emit_stdout({"ok": False, "error": "Interrupted"})
        return 130
    except Exception as exc:
        _log(f"error: {exc}")
        _emit_stdout({"ok": False, "error": str(exc)})
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: vm-agent/src/lib/mineru-upload.ts
/**
 * MinerU batch API client — TypeScript port of mineru_upload.py.
 * Uploads a local file → polls extraction → downloads result zip → extracts main markdown.
 * Zero Python dependency.
 */

import { readFileSync, writeFileSync, existsSync, mkdirSync, copyFileSync, readdirSync, statSync } from "node:fs";
import { resolve, basename, extname, join, relative, posix } from "node:path";
import { tmpdir } from "node:os";
import JSZip from "jszip";

// ─── Types ──────────────────────────────────────────────────────

export interface MinerUOptions {
  token: string;
  outputDir: string;
  language?: string;
  timeoutMs?: number;
  apiBase?: string;
  signal?: AbortSignal;
}

export interface MinerUSuccess {
  ok: true;
  markdown_path: string;
  images_dir: string;
  chars: number;
}

export interface MinerUFailure {
  ok: false;
  error: string;
}

export type MinerUResult = MinerUSuccess | MinerUFailure;

// ─── Constants ──────────────────────────────────────────────────

const SUPPORTED_EXTENSIONS = new Set([".pdf", ".docx", ".doc", ".pptx", ".ppt", ".png", ".jpg", ".jpeg"]);
const DEFAULT_TIMEOUT_MS = 600_000;
const POLL_INTERVAL_MS = 2_000;

// ─── Helpers ────────────────────────────────────────────────────

function log(msg: string): void {
  process.stderr.write(`[mineru_upload] ${msg}\n`);
}

function sanitizeFilename(text: string): string {
  let sanitized = text.trim().replace(/[^a-zA-Z0-9._-]+/g, "_").replace(/^[._]+|[._]+$/g, "");
  if (!sanitized) return "document";
  return sanitized.slice(0, 120);
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) { reject(new Error("Aborted")); return; }
    const timer = setTimeout(resolve, ms);
    const onAbort = () => { clearTimeout(timer); reject(new Error("Aborted")); };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}

async function httpJson(
  method: string,
  url: string,
  opts: { headers?: Record<string, string>; body?: unknown; timeoutMs?: number; signal?: AbortSignal },
): Promise<Record<string, unknown>> {
  const headers: Record<string, string> = { Accept: "application/json", ...opts.headers };
  let bodyStr: string | undefined;
  if (opts.body !== undefined) {
    bodyStr = JSON.stringify(opts.body);
    headers["Content-Type"] ??= "application/json";
  }
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), opts.timeoutMs ?? 60_000);
  if (opts.signal) {
    opts.signal.addEventListener("abort", () => controller.abort(), { once: true });
  }
  try {
    const resp = await fetch(url, { method, headers, body: bodyStr, signal: controller.signal });
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(`HTTP ${resp.status} ${method} ${url}: ${text.slice(0, 800)}`);
    }
    const text = await resp.text();
    if (!text) return {};
    return JSON.parse(text) as Record<string, unknown>;
  } finally {
    clearTimeout(timer);
  }
}

// ─── API steps ──────────────────────────────────────────────────

function mustSuccess(response: Record<string, unknown>, context: string): Record<string, unknown> {
  const code = response.code;
  if (code !== undefined && code !== null && code !== 0) {
    throw new Error(`${context} failed: ${JSON.stringify(response)}`);
  }
  const data = response.data;
  if (data && typeof data === "object" && !Array.isArray(data)) {
    return data as Record<string, unknown>;
  }
  return response;
}

function firstPresignedUrl(batchData: Record<string, unknown>): string {
  // v4 API: file_urls: ["https://..."]
  const fileUrls = batchData.file_urls;
  if (Array.isArray(fileUrls) && fileUrls.length > 0) {
    const first = fileUrls[0];
    if (typeof first === "string" && first) return first;
  }

  // Fallback: files: [{presigned_url: "..."}]
  const files = batchData.files;
  if (Array.isArray(files) && files.length > 0) {
    const first = files[0];
    if (typeof first === "string" && first) return first;
    if (first && typeof first === "object") {
      for (const key of ["presigned_url", "upload_url", "url"] as const) {
        const value = (first as Record<string, unknown>)[key];
        if (typeof value === "string" && value) return value;
      }
    }
  }

  throw new Error(`Batch response missing presigned/file URL: ${JSON.stringify(batchData)}`);
}

async function createBatch(
  apiBase: string,
  token: string,
  fileName: string,
  language: string,
  timeoutMs: number,
  signal?: AbortSignal,
): Promise<{ batchId: string; presignedUrl: string }> {
  const endpoint = apiBase.replace(/\/+$/, "") + "/api/v4/file-urls/batch";
  const response = await httpJson("POST", endpoint, {
    headers: { Authorization: `Bearer ${token}` },
    body: {
      enable_formula: true,
      enable_table: true,
      language,
      files: [{ name: fileName, is_ocr: true }],
    },
    timeoutMs: Math.min(timeoutMs, 120_000),
    signal,
  });
  const data = mustSuccess(response, "create batch");

  const batchId = (data.batch_id ?? response.batch_id) as string | undefined;
  if (!batchId) throw new Error(`Batch response missing batch_id: ${JSON.stringify(response)}`);

  const presignedUrl = firstPresignedUrl(data);
  return { batchId, presignedUrl };
}

async function uploadFile(presignedUrl: string, filePath: string, timeoutMs: number, signal?: AbortSignal): Promise<void> {
  const fileBytes = readFileSync(filePath);
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  if (signal) signal.addEventListener("abort", () => controller.abort(), { once: true });
  try {
    // OSS presigned URLs are signed without Content-Type — do NOT send one
    const resp = await fetch(presignedUrl, {
      method: "PUT",
      body: fileBytes,
      signal: controller.signal,
      // @ts-expect-error duplex required for streaming body in some runtimes
      duplex: "half",
    });
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(`HTTP ${resp.status} PUT upload: ${text.slice(0, 800)}`);
    }
  } finally {
    clearTimeout(timer);
  }
}

async function pollBatch(
  apiBase: string,
  token: string,
  batchId: string,
  timeoutMs: number,
  signal?: AbortSignal,
): Promise<{ extractResult: Record<string, unknown>; zipUrl: string }> {
  const endpoint = apiBase.replace(/\/+$/, "") + `/api/v4/extract-results/batch/${batchId}`;
  const start = Date.now();
  let lastState: string | null = null;

  while (true) {
    const response = await httpJson("GET", endpoint, {
      headers: { Authorization: `Bearer ${token}` },
      timeoutMs: Math.min(timeoutMs, 60_000),
      signal,
    });
    const data = mustSuccess(response, "poll batch");

    // Resolve state
    let state = (data.state ?? response.state) as string | undefined;

    const extractResult =
      (data.extract_result ?? data.extract_results ?? response.extract_result ?? response.extract_results) as
        | Record<string, unknown>
        | Record<string, unknown>[]
        | undefined;

    if (!state && Array.isArray(extractResult) && extractResult.length > 0) {
      const itemStates = extractResult
        .filter((item): item is Record<string, unknown> => typeof item === "object" && item !== null)
        .map((item) => item.state as string | undefined)
        .filter(Boolean) as string[];
      if (itemStates.every((s) => s === "done")) state = "done";
      else if (itemStates.some((s) => s === "failed")) state = "failed";
      else if (itemStates.length > 0) state = itemStates[0];
    }

    if (state && state !== lastState) {
      const elapsed = ((Date.now() - start) / 1000).toFixed(1);
      log(`state=${state} elapsed=${elapsed}s`);
      lastState = state;
    }

    if (state === "done") {
      const items: Record<string, unknown>[] = Array.isArray(extractResult)
        ? extractResult.filter((item): item is Record<string, unknown> => typeof item === "object" && item !== null)
        : extractResult && typeof extractResult === "object"
          ? [extractResult]
          : [];

      if (items.length === 0) {
        throw new Error(`Batch done but missing extract_result: ${JSON.stringify(response)}`);
      }

      const chosen = items.find((item) => typeof item.full_zip_url === "string" && item.full_zip_url) ?? items[0];
      const zipUrl = chosen.full_zip_url as string | undefined;
      if (!zipUrl) throw new Error(`Batch done but full_zip_url missing: ${JSON.stringify(chosen)}`);
      return { extractResult: chosen, zipUrl };
    }

    if (state === "failed") {
      const errMsg =
        (data.err_msg ?? response.err_msg ?? data.message ?? response.message) as string | undefined;
      throw new Error(`MinerU batch failed: ${errMsg || "(no err_msg)"}`);
    }

    if (Date.now() - start > timeoutMs) {
      throw new Error(`Poll timeout after ${timeoutMs / 1000}s (last state=${state})`);
    }

    await sleep(POLL_INTERVAL_MS, signal);
  }
}

async function downloadBytes(url: string, timeoutMs: number, signal?: AbortSignal): Promise<Buffer> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  if (signal) signal.addEventListener("abort", () => controller.abort(), { once: true });
  try {
    const resp = await fetch(url, { signal: controller.signal });
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(`HTTP ${resp.status} GET ${url}: ${text.slice(0, 800)}`);
    }
    return Buffer.from(await resp.arrayBuffer());
  } finally {
    clearTimeout(timer);
  }
}

// ─── Zip extraction ─────────────────────────────────────────────

function walkDir(dir: string): string[] {
  const results: string[] = [];
  for (const entry of readdirSync(dir, { withFileTypes: true })) {
    const full = join(dir, entry.name);
    if (entry.isDirectory()) results.push(...walkDir(full));
    else results.push(full);
  }
  return results;
}

async function extractZip(zipBuffer: Buffer, outDir: string): Promise<string[]> {
  mkdirSync(outDir, { recursive: true });
  const zip = await JSZip.loadAsync(zipBuffer);
  const extracted: string[] = [];

  for (const [name, entry] of Object.entries(zip.files)) {
    if (entry.dir) continue;
    const target = resolve(outDir, name);
    // Security: prevent zip slip
    if (!target.startsWith(resolve(outDir))) {
      throw new Error(`Unsafe zip path: ${name}`);
    }
    mkdirSync(resolve(target, ".."), { recursive: true });
    const content = await entry.async("nodebuffer");
    writeFileSync(target, content);
    extracted.push(target);
  }

  return extracted;
}

function markdownScore(filePath: string): [number, number] {
  const name = basename(filePath).toLowerCase();
  const rel = filePath.replace(/\\/g, "/").toLowerCase();
  let penalty = 0;
  let bonus = 0;

  if (name.includes("readme")) penalty += 3;
  if (/layout|span|debug/.test(rel)) penalty += 3;
  if (/middle|origin/.test(rel)) penalty += 2;
  if (/content|result|output/.test(name)) bonus += 2;

  const size = existsSync(filePath) ? statSync(filePath).size : 0;
  return [bonus - penalty, size];
}

function findMainMarkdown(files: string[]): string {
  const candidates = files.filter((f) => {
    const ext = extname(f).toLowerCase();
    return ext === ".md" || ext === ".markdown";
  });
  if (candidates.length === 0) throw new Error("No markdown file found in MinerU result zip");

  candidates.sort((a, b) => {
    const [aScore, aSize] = markdownScore(a);
    const [bScore, bSize] = markdownScore(b);
    if (bScore !== aScore) return bScore - aScore;
    return bSize - aSize;
  });

  return candidates[0];
}

function copyImages(extractRoot: string, imagesOutDir: string): number {
  let copied = 0;

  function findImageDirs(dir: string): string[] {
    const dirs: string[] = [];
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        if (entry.name.toLowerCase() === "images") dirs.push(full);
        dirs.push(...findImageDirs(full));
      }
    }
    return dirs;
  }

  const imageDirs = findImageDirs(extractRoot);
  if (imageDirs.length === 0) return 0;

  mkdirSync(imagesOutDir, { recursive: true });
  for (const srcDir of imageDirs) {
    for (const file of walkDir(srcDir)) {
      const rel = relative(srcDir, file);
      const dst = join(imagesOutDir, rel);
      mkdirSync(resolve(dst, ".."), { recursive: true });
      copyFileSync(file, dst);
      copied++;
    }
  }
  return copied;
}

// ─── Main export ────────────────────────────────────────────────

export async function mineruUpload(filePath: string, opts: MinerUOptions): Promise<MinerUResult> {
  const startMs = Date.now();
  const absPath = resolve(filePath);
  const outputDir = resolve(opts.outputDir);
  const apiBase = opts.apiBase ?? "https://mineru.net";
  const language = opts.language ?? "ch";
  const timeoutMs = opts.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const signal = opts.signal;

  if (!existsSync(absPath)) {
    return { ok: false, error: `Input file not found: ${absPath}` };
  }

  const ext = extname(absPath).toLowerCase();
  if (!SUPPORTED_EXTENSIONS.has(ext)) {
    const supported = [...SUPPORTED_EXTENSIONS].map((e) => e.slice(1)).sort().join(", ");
    return { ok: false, error: `Unsupported file format '${ext}'. Supported: ${supported}` };
  }

  try {
    mkdirSync(outputDir, { recursive: true });
    log(`input=${absPath} output_dir=${outputDir}`);

    // 1. Create batch
    log("creating upload batch");
    const { batchId, presignedUrl } = await createBatch(apiBase, opts.token, basename(absPath), language, timeoutMs, signal);
    log(`batch_id=${batchId}`);

    // 2. Upload file
    log("uploading file to presigned URL");
    await uploadFile(presignedUrl, absPath, Math.min(timeoutMs, 300_000), signal);

    // 3. Poll until done
    log("polling extraction status");
    const { zipUrl } = await pollBatch(apiBase, opts.token, batchId, timeoutMs, signal);

    // 4. Download result zip
    log("downloading result zip");
    const zipBuffer = await downloadBytes(zipUrl, Math.min(timeoutMs, 300_000), signal);

    // 5. Extract zip → find main markdown + images
    const tempExtractDir = join(tmpdir(), `mineru-${sanitizeFilename(batchId)}-${Date.now()}`);
    try {
      const extractedFiles = await extractZip(zipBuffer, tempExtractDir);
      const mainMd = findMainMarkdown(extractedFiles);

      let markdownPath = join(outputDir, `${sanitizeFilename(basename(absPath, ext))}.md`);
      if (existsSync(markdownPath)) {
        markdownPath = join(outputDir, `${sanitizeFilename(basename(absPath, ext))}-${Date.now()}.md`);
      }
      copyFileSync(mainMd, markdownPath);

      const imagesDir = join(outputDir, "images");
      copyImages(tempExtractDir, imagesDir);

      const markdownText = readFileSync(markdownPath, "utf-8").trim();
      const elapsed = ((Date.now() - startMs) / 1000).toFixed(1);
      log(`completed batch_id=${batchId} chars=${markdownText.length} elapsed=${elapsed}s`);

      return {
        ok: true,
        markdown_path: markdownPath,
        images_dir: imagesDir,
        chars: markdownText.length,
      };
    } finally {
      // Clean up temp dir
      try {
        const { rmSync } = await import("node:fs");
        rmSync(tempExtractDir, { recursive: true, force: true });
      } catch { /* ignore cleanup errors */ }
    }
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    log(`error: ${msg}`);
    return { ok: false, error: msg };
  }
}
